matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/fastSLAM_func.m

```diff
@@ -0,0 +1,172 @@
+%% Fast SLAM LOOP
+for t=2:finalTime/timeInterval
+    timeStamp(t)=timeStamp(t-1)+timeInterval;
+    pose(:,t)=motionUpdate_fs(pose(:,t-1),u,Q,timeInterval);
+    meas_ind=featureSearch_fs(noFeatures,map,pose(:,t),rmax,thmax); 
+    y=[];
+    for feature_index = meas_ind       
+        y_new=measurementUpdate_fs(map(:,feature_index),pose(:,t),R);
+        y=[y,y_new];     
+    end
+
+    for particle=1:totalParticles
+       particleSetP(:,particle)=motionUpdate_fs(particleSet(:,particle),u,Q,timeInterval);
+       clear yw hmuw Qw;
+       for j=1:length(meas_ind)
+           i = meas_ind(j);
+           [muFeatPred(:,i,particle),covFeatPred(:,:,i,particle),Hmu]=...
+               ekf_fs(particleSetP(:,particle),y(:,j),muFeat(:,i,particle),...
+               covFeat(:,:,i,particle),R,newfeature(i));
+           yw(2*(j-1)+1:2*j) = y(:,j);
+           hmuw(2*(j-1)+1:2*j) = Hmu;
+           Qw(2*(j-1)+1:2*j,2*(j-1)+1:2*j) = 2*covFeatPred(:,:,i,particle);
+       end
+       if (exist('Qw'))
+            w(particle) = max(0.000001,mvnpdf(yw,hmuw,Qw));
+       end
+    end
+        newfeature(meas_ind) = 0; 
+        [muAllParticles,particleSet,muFeat,covFeat]=raoBlackwellFilter_fs(totalParticles...
+            ,w,particleSetP,muFeatPred,covFeatPred);
+        centroid_particles(:,t) = muAllParticles;
+    
+     plot_fs(pose,map,y,muFeat,particleSet,t,meas_ind,totalParticles,noFeatures,newfeature,centroid_particles);
+       
+end
+
+
+    %% Individual Function Definations
+    %% Initialization Function   
+    function[timeInterval,finalTime,noFeatures,map,pose,totalParticles,muFeat,muFeatPred,covFeat,covFeatPred,newfeature,w,u,timeStamp,rmax,thmax]...
+            =initializeFastSLAM(PoseInitial,MotionNoise,MeasurementNoise,MapSize,FinalTime,dt,ParticleSize)
+        poseInitial=PoseInitial;
+        timeInterval=dt;
+        finalTime=FinalTime;%final Time in sec
+        noFeatures=MapSize;
+        map = 10*rand(2,noFeatures);
+        map(1,:) = map(1,:)-5; 
+        map(2,:) = map(2,:)-2; 
+        Q=MeasurementNoise;R=MotionNoise;
+        rmax = 10;
+        thmax = pi/4;
+        pose(:,1)=poseInitial;
+        totalParticles=ParticleSize;
+        particleSet=zeros(size(poseInitial,1),totalParticles);
+        measStateSize=size(R,1);
+        poseSize=size(Q,1);
+        muFeat=zeros(measStateSize,noFeatures,totalParticles);
+        muFeatPred=muFeat;
+        covFeat=zeros(measStateSize,measStateSize,noFeatures,totalParticles);
+        covFeatPred=covFeat;
+        newfeature = ones(totalParticles,1);
+        w_initial=1/totalParticles;
+        w=w_initial*ones(1,totalParticles);
+        u=[1;0.3];
+        timeStamp(1)=timeInterval; 
+    end
+
+    %% Motion Model
+    function newPose=motionUpdate_fs(oldPose,controlInput,MotionNoise,dt)
+    [QE, Qe] = eig(MotionNoise);n = length(MotionNoise(:,1)); 
+     e = QE*sqrt(Qe)*randn(n,1);
+    % Update robot state
```

Comment:
If the motion model can also be the argument of fastSLAM(), this program is lovely:)